packages/noya-state/src/reducers/application.ts

```diff
@@ -249,59 +258,147 @@ export function reducer(
         state.selectedPage = action[1];
       });
     }
-    case 'alignLeft': {
+    case 'distribute': {
       const pageIndex = getCurrentPageIndex(state);
       const layerIndexPaths = getSelectedLayerIndexPaths(state);
+      const axis = action[1];
 
       return produce(state, (state) => {
+        const page = state.sketch.pages[pageIndex];
+        const selectedRect = getSelectedRect(layerIndexPaths, page);
         const layers = accessPageLayers(state, pageIndex, layerIndexPaths);
-        const [leftmostLayer] = layers.sort(
-          (layerA, layerB) => layerA.frame.x - layerB.frame.x,
+        const combinedWidths = layers.reduce(
+          (width, layer) => layer.frame.width + width,
+          0,
         );
-        layers.forEach((layer) => {
-          layer.frame.x = leftmostLayer.frame.x;
-        });
-      });
-    }
-    case 'alignRight': {
-      const pageIndex = getCurrentPageIndex(state);
-      const layerIndexPaths = getSelectedLayerIndexPaths(state);
-
-      return produce(state, (state) => {
-        const layers = accessPageLayers(state, pageIndex, layerIndexPaths);
-        const [rightmostLayer] = layers.sort(
-          (layerA, layerB) => layerB.frame.x - layerA.frame.x,
+        const combinedHeights = layers.reduce(
+          (height, layer) => layer.frame.height + height,
+          0,
         );
-        layers.forEach((layer) => {
-          layer.frame.x = rightmostLayer.frame.x;
-        });
-      });
-    }
-    case 'alignTop': {
-      const pageIndex = getCurrentPageIndex(state);
-      const layerIndexPaths = getSelectedLayerIndexPaths(state);
-
-      return produce(state, (state) => {
-        const layers = accessPageLayers(state, pageIndex, layerIndexPaths);
-        const [topmostLayer] = layers.sort(
-          (layerA, layerB) => layerA.frame.y - layerB.frame.y,
-        );
-        layers.forEach((layer) => {
-          layer.frame.y = topmostLayer.frame.y;
+        const differenceWidth = selectedRect.width - combinedWidths;
+        const differenceHeight = selectedRect.height - combinedHeights;
+        const gapX = differenceWidth / (layers.length - 1);
+        const gapY = differenceHeight / (layers.length - 1);
+        const sortBy = axis === 'horizontal' ? 'minX' : 'minY'; // TODO: sort by mid point once it's available
+        const sortedLayerIndexPaths = layerIndexPaths
```

Comment:
The input to this `map(...).sort(...).map(...)` operation is type `IndexPath[]`, and the output is also `IndexPath[]`. I'm wondering if we can get away with just a `sort(...)` and no maps?

```ts
const getNormalizedBounds = (indexPath: IndexPath): Bounds => {
  const layer = Layers.access(page, layerIndexPath);
  const transform = getNormalizedTransform(page, layerIndexPath);
  return Primitives.createBounds(
    Primitives.transformRect(layer.frame, transform)
  )
}

const sortedLayerIndexPaths = layerIndexPaths
  .sort((a, b) => getNormalizedBounds(a)[sortBy] - getNormalizedBounds(b)[sortBy]);
```

We also probably want to use the `Primitives.transformRect` helper for applying a transform to a layer. Otherwise we won't transform the width/height correctly (which doesn't happen yet, but I think if a symbol is scaled?). 

We could also make a `Primitives.transformBounds` to avoid a small amount of extra work here, but not too important

We could always memoize `getBounds` if needed for perf, but also probably not needed

Reply:
So much nicer now! ✨